docs/developer/cluster-lifecycle-api-provider-integration.md

```diff
@@ -0,0 +1,60 @@
+
+#### Immutable objects
+
+The biggest challenge when generating CAPI specs is the generation of new objects for immutable CRDs. The current pattern the CLI implements is: retrieving the current state of eks-a CRs and based on the diff with the new desired spec and determine if such immutable objects would change after executing the conversion logic or not. This comparison is mostly and implicit version of the reverse transformation logic, one that tries to reverse the eks-a -> CAPI conversion logic.
+
+Unfortunately this is not possible in the controller world since there is no old and new version of objects, only current. To solve this, the original cluster implements a reverse logic, trying to infer the past state of eks-a objects from the current state of CAPI objects. Then it plugs into the main CLI logic, running the previously mentioned comparison logic. This is extremely prone to errors and very hard to maintain.
+
+We encourage all providers to follow the reverse algorithm: instead of comparing the old and new inputs and inferring if the outputs would differ, compare directly the old and new output. This is a way simpler process: retrieve the current status of the immutable CAPI object, run the conversion logic to generate that same object based on the current EKS-A spec and just compare them.
+
+When used in combination with API structs, the comparison step becomes trivial most of the times, being able to take advantage of `equality.Semantic.DeepDerivative` provided by the `apimachinery` module. Examples of this can be found in the [Snow](https://github.com/aws/eks-anywhere/blob/main/pkg/providers/snow/objects.go) provider, which reuses the same code/logic for both the controller and the CLI.
+
+## Tooling
+
+The are some packages that can be useful when implementing a `ProviderClusterReconciler`. Their use is encouraged but not enforced.
+
+* [`PhaseRunner`](https://github.com/aws/eks-anywhere/blob/main/pkg/controller/runner.go). Useful for running reconciliation logic that can divided in phases where each phase can abort the execution by either erroring out or requesting a reconcile request requeue (for example, to wait some time and so it check again for a particular event).
+* [Helpers to retrieve CAPI objects](https://github.com/aws/eks-anywhere/blob/main/pkg/controller/clusterapi.go)
+* [Helpers to use server-side apply](https://github.com/aws/eks-anywhere/tree/main/pkg/controller/serverside)
+* [Helpers to extract data from CAPI objects](https://github.com/aws/eks-anywhere/blob/main/pkg/controller/clusters/clusterapi.go)
+* [Build cluster.Spec from a client using `cluster.BuildSpec`](https://github.com/aws/eks-anywhere/blob/main/pkg/cluster/fetch.go#L159)
+* [`ObjectApplier`](https://github.com/aws/eks-anywhere/blob/main/pkg/controller/serverside/applier.go#L16): helpful when generating the CAPI spec as API structs (`client.Object`s), allowing to easily reconcile them using server-side apply.
```

Comment:
This may be a larger issue, but I think we're missing some details on how debug the code with this model. I found it to be very valuable to use the GoLand debugger when observing issues with cluster creation, for example. What would be the equivalent approach when all operations are performed on the EKS-A controller? Is the expectation to use Tilt with a debugger? If so, are there any instructions for setting that up?

Reply:
That's a good topic. However, I am of the opinion that tooling should be built when it's needed.

It's definitely possible to run tilt with a debugger, but we haven't had the need yet. And yes, I would say that's probably the right solution to the problem you are describing. We will just need to build it when we get there.